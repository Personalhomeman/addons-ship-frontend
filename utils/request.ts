import fetch, { RequestInit, Response } from 'node-fetch';
import getConfig from 'next/config';

import { RequestError } from '@/models/errors';

const {
  publicRuntimeConfig: { VERBOSE_LOGGING }
} = getConfig();

const verboseLogging = Boolean(VERBOSE_LOGGING);

const getHeaders = (token: string) => ({
  Authorization: token
});

interface RequestParams extends RequestInit {
  url: string;
  token?: string;
}

export const request = ({ url, method, token, body }: RequestParams) =>
  fetch(url, {
    method,
    headers: { ...(token ? getHeaders(token) : undefined), Accept: 'application/json' },
    body
  }).then((res: Response) => {
    if (verboseLogging) {
      console.log('Request', { url, method, token, body });
    }

    if (!res.ok) {
      // throw new Error(`${res.status}: ${res.statusText}`);
      throw new RequestError(res.status, res.statusText);
    }

    return res;
  });

export const get = (url: string, token: string) => request({ method: 'GET', url, token });
export const del = (url: string, token: string) => request({ method: 'DELETE', url, token });
export const patch = (url: string, token: string, body?: RequestInit['body']) =>
  request({ method: 'PATCH', url, token, body });
export const post = (url: string, token: string, body: RequestInit['body']) =>
  request({ method: 'POST', url, token, body });
export const put = (url: string, token: string, body: RequestInit['body']) =>
  request({ method: 'PUT', url, token, body });
