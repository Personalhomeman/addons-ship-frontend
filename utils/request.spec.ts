jest.mock('node-fetch');

import fetch from 'node-fetch';

import * as request from './request';

describe('request', () => {
  const methodsWithoutBody = ['get', 'patch', 'del'],
    methodsWithBody = ['post', 'put', 'patch'],
    url = 'some.url',
    token = 'a-token',
    body = 'some-body';

  methodsWithoutBody.forEach(method => {
    test(method, async () => {
      ((fetch as unknown) as jest.Mock).mockResolvedValueOnce({ ok: true });
      await request[method](url, token);

      expect(fetch).toHaveBeenCalledWith(url, {
        method: (method === 'del' ? 'delete' : method).toUpperCase(),
        headers: {
          Authorization: token,
          Accept: 'application/json'
        }
      });
    });
  });

  methodsWithBody.forEach(method => {
    test(method, async () => {
      ((fetch as unknown) as jest.Mock).mockResolvedValueOnce({ ok: true });
      await request[method](url, token, body);

      expect(fetch).toHaveBeenCalledWith(url, {
        method: method.toUpperCase(),
        headers: {
          Authorization: token,
          Accept: 'application/json'
        },
        body
      });
    });
  });

  it('should throw an error for a bad request', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValueOnce({
      ok: false,
      status: 500,
      statusText: 'Internal Server Error'
    });

    await expect(request.get('some-url', 'a-token')).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should only include auth header when token was set', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValueOnce({ ok: true });

    const url = 'some.url',
      method = 'GET';
    await request.request({ url, method });

    expect(fetch).toHaveBeenCalledWith(url, { method, headers: { Accept: 'application/json' } });
  });
});
