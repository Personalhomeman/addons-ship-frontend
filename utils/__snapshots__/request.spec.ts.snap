// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request should throw an error for a bad request 1`] = `"Cannot set property 'status' of undefined"`;
